src/NUnitFramework/framework/Constraints/CollectionTally.cs

```diff
@@ -104,15 +104,17 @@ private bool ItemsEqual(object expected, object actual)
             return comparer.AreEqual(expected, actual, ref tolerance);
         }
 
-        private static bool TrySort(ArrayList items)
+        private static bool TrySort(ref ArrayList items)
         {
+            var original = (ArrayList)items.Clone();
```

Comment:
I wonder if it would be cheaper to check if all elements are IComparable iso making a Clone.

Reply:
Hi @manfred-brands !
I'm not terribly happy with this clone approach myself. This is a bit of a tricky edge case where, in the case of a `Tuple<T1, T2, etc>` the tuple itself may implement IComparable but will throw an exception when you try to call it. This is because the tuple's properties must all implement IComparable for the tuple itself to be sortable but the Tuple doesn't actually perform this check until runtime. Rather than for us to do this type check recursively on each item, a decision was made in https://github.com/nunit/nunit/pull/3831 to assume the happypath and catch the exception and fallback if it can't be sorted.

The thorny part of this, however, is that the in-place Array.Sort() sorting algorithm used by ArrayList + List<T> seems to leave the array in an inconsistent state if an exception is thrown in this case in .NET 4.5+. I'm still working through a few options but it is looking like the simplest option is something which avoids using Array.Sort() either directly or indirectly.

Reply:
Thanks for the explanation.
This ticket seems to have stalled, even if not ideal, this at least would fix the issue were are currently seeing after upgrading to 3.12.2